File: advanced-programming-methods/labs/toy-interpreter-sharp/model/expression/ArithmeticExpression.cs
using System;
using Exceptions;

namespace Model
{
    /// <summary>
    /// The type Arithmetic expchression.
    /// </summary>
    [Serializable()]
    public class ArithmeticExpression : Expression
    {
        private readonly Expression left;
        private readonly Expression right;
        private readonly int @operator;

        /// <summary>
        /// Instantiates a new Arithmetic expression.
        /// </summary>
        /// <param name="left"> the left </param>
        /// <param name="right"> the right </param>
        /// <param name="operator"> the operator </param>
        public ArithmeticExpression(Expression left, Expression right, int @operator)
        {
            this.left = left;
            this.right = right;
            this.@operator = @operator;
        }

        /// <inheritdoc />
        public override int Eval(IDictionary table, IHeap heap)
        {
            int evalLeft = left.Eval(table, heap);
            int evalRight = right.Eval(table, heap);

            switch (@operator)
            {
                case Operation.ADD:
                    return evalLeft + evalRight;

                case Operation.SUBTRACT:
                    return evalLeft - evalRight;

                case Operation.DIVIDE:
                    return evalLeft / evalRight;

                case Operation.MULTIPLY:
                    return evalLeft * evalRight;

                default:
                    throw new ExpressionException("The operator is invalid");
            }
        }

        /// <inheritdoc />
        public override string ToString()
        {
            return string.Format("{0} {1} {2}", left, Operation.OperationToString(@operator), right);
        }

        /// <inheritdoc />
        public override Expression CloneDeep()
        {
            return new ArithmeticExpression(left.CloneDeep(), right.CloneDeep(), @operator);
        }
    }
}

File: advanced-programming-methods/labs/toy-interpreter-java/src/model/expression/ArithmeticExpression.java
package model.expression;

import exceptions.ExpressionException;
import model.IDictionary;
import model.IHeap;

/**
 * The type Arithmetic expression.
 */
public class ArithmeticExpression extends Expression
{
    private Expression left;
    private Expression right;
    private int        operator;

    /**
     * Instantiates a new Arithmetic expression.
     *
     * @param left     the left
     * @param right    the right
     * @param operator the operator
     */
    public ArithmeticExpression(Expression left, Expression right, int operator)
    {
        this.left = left;
        this.right = right;
        this.operator = operator;
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public int eval(IDictionary table, IHeap heap) throws ExpressionException
    {
        int evalLeft = left.eval(table, heap);
        int evalRight = right.eval(table, heap);

        switch (operator)
        {
            case Operation.ADD:
                return evalLeft + evalRight;

            case Operation.SUBTRACT:
                return evalLeft - evalRight;

            case Operation.DIVIDE:
                return evalLeft / evalRight;

            case Operation.MULTIPLY:
                return evalLeft * evalRight;

            default:
                throw new ExpressionException("The operator is invalid");
        }
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public String toString()
    {
        return String.format("%s %s %s", left.toString(), Operation.operationToString(operator), right.toString());
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public Expression cloneDeep()
    {
        return new ArithmeticExpression(left.cloneDeep(), right.cloneDeep(), operator);
    }
}
